File: airbyte-integrations/bases/base-java-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/AvroNameTransformer.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.s3.avro;

import io.airbyte.integrations.destination.ExtendedNameTransformer;
import java.util.Arrays;

/**
 * <ul>
 * <li>An Avro name starts with [A-Za-z_], followed by [A-Za-z0-9_].</li>
 * <li>An Avro namespace is a dot-separated sequence of such names.</li>
 * <li>Reference: https://avro.apache.org/docs/current/spec.html#names</li>
 * </ul>
 */
public class AvroNameTransformer extends ExtendedNameTransformer {

  @Override
  public String applyDefaultCase(final String input) {
    return super.convertStreamName(input).toLowerCase();
  }

  @Override
  public String convertStreamName(final String input) {
    if (input == null) {
      return null;
    } else if (input.isBlank()) {
      return input;
    }

    final String normalizedName = super.convertStreamName(input);
    if (normalizedName.substring(0, 1).matches("[A-Za-z_]")) {
      return normalizedName;
    } else {
      return "_" + normalizedName;
    }
  }

  @Override
  public String getNamespace(final String input) {
    if (input == null) {
      return null;
    }

    final String[] tokens = input.split("\\.");
    return String.join(".", Arrays.stream(tokens).map(this::getIdentifier).toList());
  }

}

File: airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/avro/AvroNameTransformer.kt
/*
 * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.destination.s3.avro

import io.airbyte.cdk.integrations.destination.StandardNameTransformer
import java.util.Locale

/**
 *
 * * An Avro name starts with [A-Za-z_], followed by [A-Za-z0-9_].
 * * An Avro namespace is a dot-separated sequence of such names.
 * * Reference: https://avro.apache.org/docs/current/spec.html#names
 */
class AvroNameTransformer : StandardNameTransformer() {
    override fun applyDefaultCase(input: String): String {
        return super.convertStreamName(input).lowercase(Locale.getDefault())
    }

    override fun convertStreamName(input: String): String {
        val normalizedName = super.convertStreamName(input)
        return if (normalizedName.substring(0, 1).matches("[A-Za-z_]".toRegex())) {
            normalizedName
        } else {
            "_$normalizedName"
        }
    }

    override fun getNamespace(namespace: String): String {
        val tokens = namespace.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return tokens
            .map { name: String ->
                this.getIdentifier(
                    name,
                )
            }
            .joinToString(separator = ".")
    }
}
